src/request/payload.rs

```diff
@@ -1,23 +1,100 @@
 use crate::error::Error;
-use crate::request::notification::{DefaultAlert, NotificationOptions, WebPushAlert};
+use crate::request::notification::{DefaultAlert, DefaultSound, NotificationOptions, WebPushAlert};
 use erased_serde::Serialize;
 use serde_json::{self, Value};
 use std::collections::BTreeMap;
 
 /// The data and options for a push notification.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize)]
 pub struct Payload<'a> {
     /// Send options
+    #[serde(skip)]
     pub options: NotificationOptions<'a>,
     /// The token for the receiving device
+    #[serde(skip)]
     pub device_token: &'a str,
     /// The pre-defined notification payload
     pub aps: APS<'a>,
     /// Application specific payload
+    #[serde(flatten)]
     pub data: BTreeMap<&'a str, Value>,
 }
 
+/// Object that can be serialized to create an APNS request.
+/// You probably just want to use [`Payload`], which implements [`PayloadLike`].
+///
+/// # Example
+/// ```no_run
+/// use a2::client::Endpoint;
+/// use a2::request::notification::{NotificationBuilder, NotificationOptions};
+/// use a2::request::payload::{PayloadLike, APS};
+/// use a2::Client;
+/// use a2::DefaultNotificationBuilder;
+/// use serde::Serialize;
+/// use std::fs::File;
+///
+/// async fn send() -> Result<(), Box<dyn std::error::Error>> {
+///     let builder = DefaultNotificationBuilder::new()
+///         .set_body("Hi there")
+///         .set_badge(420)
+///         .set_category("cat1")
+///         .set_sound("ping.flac");
+///
+///     let payload = builder.build("device-token-from-the-user", Default::default());
+///     let mut file = File::open("/path/to/private_key.p8")?;
+///
+///     let client = Client::token(&mut file, "KEY_ID", "TEAM_ID", Endpoint::Production).unwrap();
+///
+///     let response = client.send(payload).await?;
+///     println!("Sent: {:?}", response);
+///     Ok(())
+/// }
+///
+/// #[derive(Serialize)]
+/// struct Payload<'a> {
+///     aps: APS<'a>,
+///     my_custom_value: String,
+///     #[serde(skip_serializing)]
+///     options: NotificationOptions<'a>,
+///     #[serde(skip_serializing)]
+///     device_token: &'a str,
+/// }
+///
+/// impl<'a> PayloadLike for Payload<'a> {
+///     fn get_device_token(&self) -> &'a str {
+///         self.device_token
+///     }
+///     fn get_options(&self) -> &NotificationOptions {
+///         &self.options
+///     }
+/// }
+/// ```
+pub trait PayloadLike: serde::Serialize {
+    /// Combine the APS payload and the custom data to a final payload JSON.
+    /// Returns an error if serialization fails.
+    #[allow(clippy::wrong_self_convention)]
+    fn to_json_string(&self) -> Result<String, Error> {
+        Ok(serde_json::to_string(&self)?)
+    }
+
+    /// Returns token for the device
+    fn get_device_token(&self) -> &str;
+
+    /// Gets [`NotificationOptions`] for this Payload.
+    fn get_options(&self) -> &NotificationOptions;
+}
+
+impl<'a> PayloadLike for Payload<'a> {
```

Comment:
So... my company's app puts some data into the `aps` key of the payload (bad idea, we know, we're kinda stuck with it though so older versions of the app can still receive pushes). Doing it this way means we can insert our own stuff into `aps`:

```
#[derive(Serialize, Debug, Clone)]
struct APSWrapper<'a> {
  r#type: &'a str,
  #[serde(flatten)]
  aps: APS<'a>,
}

#[derive(Serialize, Debug, Clone)]
struct Payload<'a> {
  aps: APSWrapper<'a>,
  details: IOSDetails<'a>,
  #[serde(serialize_with = "bson::serde_helpers::serialize_object_id_as_hex_string")]
  id: ObjectId,
  #[serde(skip_serializing)]
  device_token: &'a str,
  #[serde(skip_serializing)]
  options: NotificationOptions<'a>,
}

impl<'a> PayloadLike for Payload<'a> {
  fn get_device_token(&self) -> &'a str {
    self.device_token
  }

  fn get_options(&self) -> &NotificationOptions {
    &self.options
  }
}
```

The other bit that's nice about this is it lets us enforce types on all the keys (since we `impl Serialize`)

This could probably be a derive macro, but I don't really know how to do that yet :stuck_out_tongue:

Reply:
I agree it would be nice to split it out into a separate PR, but the way I was writing my code (which was my smoke test) meant we needed to have this feature to make sure the whole chain worked

Reply:
Okay, I'm going to ping someone else internally to review as well then as this is a breaking change